osu.Server.Queues.ScoreStatisticsProcessor.Tests/MedalProcessorTests.cs

```diff
@@ -0,0 +1,198 @@
+using Xunit;
+
+namespace osu.Server.Queues.ScoreStatisticsProcessor.Tests
+{
+    public class MedalProcessorTests : DatabaseTest
+    {
+        private readonly List<MedalProcessor.AwardedMedal> awardedMedals = new List<MedalProcessor.AwardedMedal>();
+
+        public MedalProcessorTests()
+        {
+            MedalProcessor.MedalAwarded += onMedalAwarded;
+
+            using (var db = Processor.GetDatabaseConnection())
+            {
+                db.Execute("TRUNCATE TABLE osu_achievements");
+                db.Execute("TRUNCATE TABLE osu_user_achievements");
+
+                db.Execute("TRUNCATE TABLE osu_beatmappacks");
+                db.Execute("TRUNCATE TABLE osu_beatmappacks_items");
+            }
+        }
+
+        [Fact]
+        public void TestOnlyAwardsOnce()
+        {
+            const int medal_id = 7;
+            const int pack_id = 40;
+
+            addPackMedal(medal_id, pack_id, AllBeatmaps);
+
+            assertNotAwarded();
+            setScoreForBeatmap(TEST_BEATMAP_ID);
+
+            assertAwarded(medal_id);
+
+            setScoreForBeatmap(TEST_BEATMAP_ID);
+            assertAwarded(medal_id);
+        }
+
+        [Fact]
+        public void TestSimplePack()
+        {
+            const int medal_id = 7;
+            const int pack_id = 40;
+
+            AddBeatmap(b => b.beatmap_id = 71621, s => s.beatmapset_id = 13022);
+            AddBeatmap(b => b.beatmap_id = 59225, s => s.beatmapset_id = 16520);
+            AddBeatmap(b => b.beatmap_id = 79288, s => s.beatmapset_id = 23073);
+            AddBeatmap(b => b.beatmap_id = 101236, s => s.beatmapset_id = 27936);
+            AddBeatmap(b => b.beatmap_id = 105325, s => s.beatmapset_id = 32162);
+            AddBeatmap(b => b.beatmap_id = 127762, s => s.beatmapset_id = 40233);
+            AddBeatmap(b => b.beatmap_id = 132751, s => s.beatmapset_id = 42158);
+            AddBeatmap(b => b.beatmap_id = 134948, s => s.beatmapset_id = 42956);
+            AddBeatmap(b => b.beatmap_id = 177972, s => s.beatmapset_id = 59370);
+            AddBeatmap(b => b.beatmap_id = 204837, s => s.beatmapset_id = 71476);
+            AddBeatmap(b => b.beatmap_id = 206298, s => s.beatmapset_id = 72137);
+            AddBeatmap(b => b.beatmap_id = 271917, s => s.beatmapset_id = 102913);
+            AddBeatmap(b => b.beatmap_id = 514849, s => s.beatmapset_id = 169848);
+            AddBeatmap(b => b.beatmap_id = 497769, s => s.beatmapset_id = 211704);
+
+            addPackMedal(medal_id, pack_id, AllBeatmaps);
+
+            foreach (var beatmap in AllBeatmaps)
+            {
+                assertNotAwarded();
+                setScoreForBeatmap(beatmap.beatmap_id);
+            }
+
+            assertAwarded(medal_id);
+
+            WaitForDatabaseState("SELECT playcount FROM osu_user_stats WHERE user_id = 2", AllBeatmaps.Count, CancellationToken);
+        }
+
+        [Fact]
+        public void TestPlayMultipleBeatmapsFromSameSet()

```

Comment:
As discussed IRL, I think these names should be updated to reflect the assertion. For this case maybe something like `TestPlayMultipleBeatmapsFromSameSetDoesNotAward`.

Reply:
I've added more documentation and tidied things up. See how it reads now.